docs: section about contract negotiations (#182)
eclipse-edc/docs: developer/wip/for-contributors/contributor-handbook.md

#### 2.1.4 Contract negotiations

If a connector fulfills the [contract policy](#213-contract-definitions), it may initiate the negotiation of a contract for
a particular asset. During that negotiation, both parties can send offers and counter-offers that can contain altered
terms (= policy) as any human would in a negotiation, and the counter-party may accept or reject them.

Contract negotiations have a few key aspects:

- they target _one_ asset
- they take place between a _provider_ and a _consumer_ connector
- they cannot be changed by the user directly
- users can only decline, terminate or cancel them

As a side note it is also important to note that contract offers are _ephemeral_ objects as they are generated
on-the-fly for a particular participant, and they are never persisted in a database and thus cannot be queried through
any API.

Contract negotiations are asynchronous in nature. That means after initiating them, they become (potentially
long-running) stateful processes that are advanced by an internal [state machine](#221-state-machines).
The current state of the negotiation can be queried and altered through the management API.

> > In the current iteration all contract offers are automatically accepted by both parties if the contract

Reviewer: we already have them -> guards: https://github.com/eclipse-edc/Connector/blob/de80081d547c66f127dd32e1a8b27d9fca2d4410/docs/developer/state-machine.md#guards

```json
{
    "@context": {
        "@vocab": "https://w3id.org/edc/v0.0.1/ns/"
    },
    "@type": "ContractRequest",
    "counterPartyAddress": "http://provider-address",
    "protocol": "dataspace-protocol-http",
    "policy": {
        "@context": "http://www.w3.org/ns/odrl.jsonld",
        "@type": "odrl:Offer",
        "@id": "offer-id",
        "assigner": "providerId",
        "permission": [],
        "prohibition": [],
        "obligation": [],
        "target": "assetId"
    },
    "callbackAddresses": [
        {
            "transactional": false,
            "uri": "http://callback/url",
            "events": [
                "contract.negotiation"
            ],
            "authKey": "auth-key",
            "authCodeId": "auth-code-id"
        }
    ]
}
```
The `counterPartyAddress` is the address where to send the `ContractRequestMessage` via the specified `protocol` (currently [`dataspace-protocol-http`](#27-protocol-extensions-dsp))

The `policy` should hold the same policy associated to the data offering chosen from the [catalog](#216-catalog), plus two additional properties:

> - `assigner` the provider ID

Reviewer: ```suggestion
- `assigner` the providers `participantId`
```

- `target` the asset (dataset) ID

> In addition the `callbackAddresses`, which is optional, can be used to get notified about state changes of the negotiation. Read more on callbacks in the section about [events and callbacks](#51-events-and-callbacks).

Reviewer: ```suggestion
In addition, the (optional) `callbackAddresses` array can be used to get notified about state changes of the negotiation. Read more on callbacks in the section about [events and callbacks](#51-events-and-callbacks).
```
